api.proto

```diff
@@ -262,18 +263,28 @@ enum AckPolicy {
 
 // Message represents a message from a stream.
 message Message {
-    int64               offset        = 1 [jstype=JS_STRING];  // Monotonic message offset in the stream
-    bytes               key           = 2;  // Message key
-    bytes               value         = 3;  // Message payload
-    int64               timestamp     = 4 [jstype=JS_STRING];  // When the message was received by the broker
-    string              stream        = 5;  // Stream name message was received on
-    int32               partition     = 6;  // Stream partition message was assigned to
-    string              subject       = 7;  // NATS subject message was received on
-    string              replySubject  = 8;  // NATS reply subject
-    map<string, bytes>  headers       = 9;  // Message headers
-    string              ackInbox      = 10; // NATS subject to publish acks to
-    string              correlationId = 11; // User-supplied value to correlate acks to publishes
-    AckPolicy           ackPolicy     = 12; // Controls the behavior of acks
+    int64               offset          = 1 [jstype=JS_STRING];  // Monotonic message offset in the stream
+    bytes               key             = 2;  // Message key
+    bytes               value           = 3;  // Message payload
+    int64               timestamp       = 4 [jstype=JS_STRING];  // When the message was received by the broker
+    string              stream          = 5;  // Stream name message was received on
+    int32               partition       = 6;  // Stream partition message was assigned to
+    string              subject         = 7;  // NATS subject message was received on
+    string              replySubject    = 8;  // NATS reply subject
+    map<string, bytes>  headers         = 9;  // Message headers
+    string              ackInbox        = 10; // NATS subject to publish acks to
+    string              correlationId   = 11; // User-supplied value to correlate acks to publishes
+    AckPolicy           ackPolicy       = 12; // Controls the behavior of acks
+    int64               expectedOffset  = 13; // Provide expected offset on the partion after publish
+
+}
+
+
+// AckError represents error that is piggybacked to AckInbox due to any error that is encountered upon publishing
+// E.g: A PublishRequest may includes the expectedOffset which is outdated on partition. This publish would be refused.
+enum AckError {
+    OK = 0;
```

Comment:
My thought is `0` should be a catch-all `UNKNOWN` error, since the `AckError` on the `Ack` will be null if there is no error?

Reply:
Making `AckError` to be null in case of no error is much more intuitive.